src/bls_pop.rs

```diff
@@ -0,0 +1,155 @@
+//! ## Implementation of ProofofPossion trait for BLS keys using schnorr sginature
+//! ## TODO: I assume this can also moved to pop.rs but for now I put it separately to help reviews
+use crate::engine::{EngineBLS};
+use crate::pop::{ProofOfPossessionGenerator, ProofOfPossessionVerifier, SchnorrProof};
+
+use crate::single::{SecretKey,PublicKey};
+
+use digest::{Digest};
+
+use ark_serialize::{CanonicalSerialize};
+use ark_ec::ProjectiveCurve;
+
+use ark_ff::bytes::{FromBytes, ToBytes};
+
+use super::Message;
+
+// TODO: Delete after migration to secret key model
+// pub struct BLSSchnorrProof<E: EngineBLS> : trait B: {
+//     pub public_key : PublicKey<E>,
+//     pub proof_of_possession : SchnorrProof<E>,
+// }
+
+/// Generate Schnorr Signature for an arbitrary message using a key ment to use in BLS scheme
+trait BLSSchnorrPoPGenerator<E: EngineBLS, H: Digest> : ProofOfPossessionGenerator<E,H> {
+    /// Produce a secret witness scalar `k`, aka nonce, from hash of
+    /// H( H(s) | H(public_key)) because our key does not have the
+    /// randomness redundacy exists in EdDSA secret key.
+    fn witness_scalar(&self) -> <<E as EngineBLS>::PublicKeyGroup as ProjectiveCurve>::ScalarField;
+}
+
+impl<E: EngineBLS, H: Digest> BLSSchnorrPoPGenerator<E,H> for SecretKey<E>
+{
+    /// TODO: BROKEN NOW Switch to https://github.com/arkworks-rs/algebra/pull/164/files
+    fn witness_scalar(&self) -> <<E as EngineBLS>::PublicKeyGroup as ProjectiveCurve>::ScalarField {
+        let mut secret_key_as_bytes = vec![0;  self.into_vartime().0.serialized_size()];
+
+        let affine_public_key = self.into_public().0.into_affine();
+        let mut public_key_as_bytes = vec![0;  affine_public_key.serialized_size()];
+
+        self.into_vartime().0.serialize(&mut secret_key_as_bytes[..]).unwrap();
+        affine_public_key.serialize(&mut public_key_as_bytes[..]).unwrap();        
+        
+        let secret_key_hash = <H as Digest>::new().chain(secret_key_as_bytes);
+        let public_key_hash = <H as Digest>::new().chain(public_key_as_bytes);
+
+        let mut scalar_bytes = <H as Digest>::new().chain(secret_key_hash.finalize()).chain(public_key_hash.finalize()).finalize();
+	    let random_scalar : &mut [u8] = scalar_bytes.as_mut_slice();
+	    random_scalar[31] &= 31; // BROKEN HACK DO BOT DEPLOY
+        <<<E as EngineBLS>::PublicKeyGroup as ProjectiveCurve>::ScalarField as FromBytes>::read(&*random_scalar).unwrap()
+    }
+
+}
+
+impl<E: EngineBLS, H: Digest> ProofOfPossessionGenerator<E,H> for SecretKey<E> {
+
+    fn generate_pok(&self, message: Message) -> SchnorrProof<E> {
```

Comment:
I'm tempted to use a transcript abstraction instead of `Message` like in schnorrkel, ring-vrf, and now the apk crate.  `Message` made sense for BLS signatures message.  And maybe still does.  I do not see much point in a schnorr proof.

Reply:
We should maybe make this PoK do both G1 and G2 if we're going that route, no?  I guess that's a seperate thing that depends upon only `Engine`.  Ugh.

Reply:
Made issues #30 and #31 to follow up with these.